src/section.rs

```diff
@@ -48,249 +48,355 @@ impl Section {
-        self.relocating_in_nodes.clear();
+    /// Call this at the begining of each simulation tick to reset some internal state.
+    pub fn prepare(&mut self) {
+        self.churned = false
     }
 
-    pub fn handle_requests(&mut self, params: &Params) -> Vec<Response> {
-        let mut responses = Vec::new();
+    /// Single simulation iteration of this section.
+    /// Note: there can be multiple section ticks per network tick.
+    pub fn tick(&mut self, params: &Params) -> Vec<Action> {
+        let mut actions = Vec::new();
+        let mut relocated_in = false;
 
-        for request in mem::replace(&mut self.requests, Vec::new()) {
+        for message in mem::replace(&mut self.messages, Vec::new()) {
             debug!(
                 "{}: received {}",
                 log::prefix(&self.prefix),
-                log::message(&request)
+                log::message(&message)
             );
 
-            responses.extend(match request {
-                Request::Live(node) => self.handle_live(params, node),
-                Request::Dead(name) => self.handle_dead(params, name),
-                Request::Merge(prefix) => self.handle_merge(params, prefix),
-                Request::RelocateRequest(src, name, node_name) => {
-                    self.handle_relocate_request(params, src, name, node_name)
+            match message {
+                Message::RelocateRequest { node_name, target } => {
+                    actions.push(if relocated_in {
+                        Action::Send(Message::RelocateReject { node_name, target })
+                    } else {
+                        self.handle_relocate_request(params, node_name, target)
+                    })
                 }
-                Request::RelocateAccept(dst, node_name) => {
-                    self.handle_relocate_accept(dst, node_name)
+                Message::RelocateAccept { node_name, target } => {
+                    actions.extend(self.handle_relocate_accept(node_name, target))
                 }
-                Request::Relocate(node) => self.handle_relocate(params, node),
-                Request::RelocateReject(target, node_name) => {
-                    self.handle_relocate_reject(target, node_name)
+                Message::RelocateReject { node_name, target } => {
+                    actions.extend(self.handle_relocate_reject(params, node_name, target));
                 }
-            })
+                Message::RelocateCommit { node, .. } => {
+                    if let Some(action) = self.handle_relocate_commit(params, node) {
+                        relocated_in = true;
+                        actions.push(action);
+                    }
+                }
+                Message::RelocateCancel { node_name, .. } => self.handle_relocate_cancel(node_name),
+            }
         }
 
-        responses
+        if !self.churned && !relocated_in {
+            if self.incoming_relocations.is_empty() {
```

Comment:
just to confirm, here, the self.churned will be set to true which means a non-empty incoming_cache will block add node event for the entire network tick iteration, even that relocation is to be cancelled later on.
is this expected?

Reply:
Hm, maybe there should be two flags instead: `added` and `dropped`?

Reply:
not sure, maybe this is wanted.
better ask Spandan to confirm this.